creator-node/src/components/tracks/TranscodeDelegationManager.js

```diff
@@ -0,0 +1,436 @@
+const axios = require('axios')
+const fs = require('fs')
+const fsExtra = require('fs-extra')
+const FormData = require('form-data')
+const retry = require('async-retry')
+
+const config = require('../../config.js')
+const Utils = require('../../utils')
+const { logger: genericLogger } = require('../../logging')
+const {
+  generateTimestampAndSignatureForSPVerification
+} = require('../../apiSigning')
+
+const CREATOR_NODE_ENDPOINT = config.get('creatorNodeEndpoint')
+const DELEGATE_PRIVATE_KEY = config.get('delegatePrivateKey')
+const NUMBER_OF_SPS_FOR_HANDOFF_TRACK = 3
+
+class TranscodeDelegationManager {
+  static async handOff({ logContext }, req) {
+    const decisionTree = { state: null }
+    const libs = req.libs
+    const logger = genericLogger.child(logContext)
+    let resp = {}
+
+    let sps
+    try {
+      decisionTree.state = 'SELECTING_RANDOM_SPS'
+      sps = await TranscodeDelegationManager.selectRandomSPs(libs)
+    } catch (e) {
+      logger.warn(`Could not select random SPs: ${e.message}`)
+      return resp
+    }
+
+    for (const sp of sps) {
+      decisionTree.sp = sp
+      try {
+        logger.info(`Handing track off to sp=${sp}`)
+        decisionTree.state = 'HANDING_OFF_TO_SP'
+
+        const transcodeAndSegmentUUID =
+          await TranscodeDelegationManager.handOffToSp(logger, { sp, req })
+
+        decisionTree.state = 'POLLING_FOR_TRANSCODE'
+        const polledTranscodeResponse =
+          await TranscodeDelegationManager.pollForTranscode(logger, {
+            sp,
+            uuid: transcodeAndSegmentUUID
+          })
+
+        decisionTree.state = 'FETCHING_FILES_AND_WRITING_TO_FS'
+        const localFilePaths =
+          await TranscodeDelegationManager.fetchFilesAndWriteToFs(logger, {
+            ...polledTranscodeResponse,
+            sp,
+            fileNameNoExtension: req.fileNameNoExtension
+          })
+
+        resp = localFilePaths
+
+        // If any of these fields are not present, throw and retry onto the next sp
+        if (
+          !resp.transcodeFilePath ||
+          !resp.segmentFileNames ||
+          !resp.m3u8FilePath
+        ) {
+          throw new Error(
+            `Missing fields from transcode fetching response. Actual response=${JSON.stringify(
+              resp
+            )}`
+          )
+        }
+
+        // If the above asserts passed, break out of the loop
+        break
+      } catch (e) {
+        // TODO: delete tmp dir in external SP if fails and continue
```

Comment:
TODO part of this task or later? we should add context or make a linear or something, TODO comments v easy to lose

Reply:
all good, I just handled in the code instead

Reply:
updated to remove files